[DX-1015,DX-1036] Create Tyk Charts Product Stack Section (Creates new helm chart structure in Product Stack section) (#4076)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-charts/tyk-stack-chart.md

```bash
helm install tyk-stack tyk-helm/tyk-stack -n tyk --create-namespace -f values.yaml
```

## Uninstalling The Chart

```bash
helm uninstall tyk-stack -n tyk
```
This removes all the Kubernetes components associated with the chart and deletes the release.

## Upgrading Chart

```bash
helm upgrade tyk-stack tyk-helm/tyk-stack -n tyk -f values.yaml
```

_Note: Migrating from tyk-pro chart_

If you were using `tyk-pro` chart for existing release, you cannot upgrade directly. Please modify the values.yaml base on your requirements and install using the new `tyk-stack` chart.

## Configuration

To get all configurable options with detailed comments:

```bash
helm show values tyk-helm/tyk-stack > values.yaml
```

You can update any value in your local `values.yaml` file and use `-f [filename]` flag to override default values during installation. 
Alternatively, you can use `--set` flag to set it in Tyk installation. See [Using Helm](https://helm.sh/docs/intro/using_helm/) for examples.

### Set Redis Connection Details (Required)

Tyk uses Redis for distributed rate-limiting and token storage. You may set `global.redis.addr` and `global.redis.pass` with redis connection 

> string and password respectively.

Author: move this into the line above

If you do not already have Redis installed, you may use these charts provided by Bitnami

```bash
helm repo add bitnami https://charts.bitnami.com/bitnami
helm install tyk-redis bitnami/redis -n tyk --create-namespace --set image.tag=6.2.13
```

Follow the notes from the installation output to get connection details and password. The DNS name of your Redis as set by Bitnami is 

> `tyk-redis-master.tyk.svc:6379` (Tyk needs the name including the port) 

Author: move this into the line above

### Set Mongo or PostgresSQL Connection Details (Required)
If you have already installed MongoDB or PostgreSQL, you can set the connection details in `global.mongo` and `global.postgres` section of values file respectively.

If not, you can use these rather excellent charts provided by Bitnami to install MongoDB or PostgreSQL:

**Mongo Installation**

```bash
helm install tyk-mongo bitnami/mongodb --version {HELM_CHART_VERSION} --set "replicaSet.enabled=true" -n tyk
```

**PostgresSQL Installation**
```bash
helm install tyk-postgres bitnami/postgresql --set "auth.database=tyk_analytics" -n tyk
```

Follow the notes from the installation output to get connection details.

>NOTE: Please make sure you are installing MongoDB or PostgreSQL versions that are supported by Tyk. Please refer to Tyk docs to get list of [supported versions]({{< ref "tyk-dashboard/database-options" >}}).

### Protect Confidential Fields with Kubernetes Secrets

In the `values.yaml` file, some fields are considered confidential, such as `APISecret`, connection strings, etc.
Declaring values for such fields as plain text might not be desired for all use cases. Instead, for certain fields,

> Kubernetes secrets can be referenced, and the chart will 

Author: move this line into the sentence above

[define container environment variables using Secret data](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/#define-container-environment-variables-using-secret-data).

This section describes how to use Kubernetes secrets to declare confidential fields.

#### Tyk Dashboard Admin

> If Tyk Dashboard bootstrapping is enabled, Tyk Dashboard admin user will be created according to the `global.adminUser` field.

Author: ```suggestion
If Tyk Dashboard bootstrapping is enabled, a Tyk Dashboard admin user will be created according to the `global.adminUser` field.
```

All admin credentials can also be set through Kubernetes secret.

> [!NOTE]
> Once `global.adminUser.useSecretName` is declared, it takes precedence over `global.adminUser.firstName`, 
> `global.adminUser.lastName`, `global.adminUser.email` and `global.adminUser.password`.

> [!WARNING]
> If `global.adminUser.useSecretName` is in use, please add all keys mentioned below to the secret.

##### Admin First Name

It can be configured via `global.adminUser.firstName` as a plain text or Kubernetes secret which includes `adminUserFirstName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Last Name

It can be configured via `global.adminUser.lastName` as a plain text or Kubernetes secret which includes `adminUserLastName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Email

It can be configured via `global.adminUser.email` as a plain text or Kubernetes secret which includes `adminUserEmail` key

> in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

Author: move this line into the line above

##### Admin Password

It can be configured via `global.adminUser.password` as a plain text or Kubernetes secret which includes `adminUserPassword` key

> in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

Author: move this line into the line above

#### APISecret

The `global.secrets.APISecret` field configures a [header value]({{< ref "tyk-oss-gateway/configuration#secret" >}}) used in every interaction with Tyk Gateway API.

It can be configured via `global.secrets.APISecret` as a plain text or Kubernetes secret which includes `APISecret` key

> in it. Then, this secret must be referenced via `global.secrets.useSecretName`.

Author: move this line into the line above

```yaml
global:
    secrets:
        APISecret: CHANGEME
        useSecretName: "mysecret" # where mysecret includes `APISecret` key with the desired value.
```

#### AdminSecret

The `global.secrets.AdminSecret` field sets a [secret]({{< ref "tyk-dashboard/configuration#admin_secret" >}}) for Admin API.

It can be configured via `global.secrets.AdminSecret` as a plain text or Kubernetes secret which includes `AdminSecret` 

> key in it. Then, this secret must be referenced via `global.secrets.useSecretName`.

Author: move this line into the line above

```yaml
global:
    secrets:
        useSecretName: "mysecret" # where mysecret includes `useSecretName` key with the desired value.
```

> > [!NOTE]

Author: use the note shortcode here

> Once `global.secrets.useSecretName` is declared, it takes precedence over `global.secrets.APISecret` and `global.secrets.AdminSecret`.

#### Dashboard License

In order to refer Tyk Dashboard license through Kubernetes secret, please use `global.secrets.useSecretName`, where 
the secret should contain a key called `DashLicense`.

#### Tyk Enterprise Developer Portal License

In order to refer Tyk Enterprise Developer Portal license through Kubernetes secret, please use 
`tyk-dev-portal.useSecretName`, where the secret should contain a key called `DevPortalLicense`.

#### Tyk Enterprise Developer Portal Admin Password

In order to refer Tyk Enterprise Developer Portal's admin password through Kubernetes secret, 
please use `global.adminUser.useSecretName`, where the secret should contain a key called `adminUserPassword`.

#### Tyk Enterprise Developer Portal Storage Connection String

In order to refer Tyk Enterprise Developer Portal connection string to the selected database through Kubernetes secret,

> please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 

Author: move this line into the line above

`DevPortalStorageConnectionString`.

> [!WARNING]
> If `tyk-dev-portal.useSecretName` is in use, please add all keys mentioned to the secret. 

#### Tyk Enterprise Developer Portal AWS S3 Access Key ID

> In order to refer Tyk Enterprise Developer Portal AWS S3 Access Key ID through Kubernetes secret,

Author: Reafctor to read:

_In order to refer to the *Tyk Enterprise Developer Portal AWS S3 Access Key ID* using a Kubernetes secret, please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 
`DevPortalAwsAccessKeyId`._

please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 
`DevPortalAwsAccessKeyId`.

> [!WARNING]
> If `tyk-dev-portal.useSecretName` is in use, please add all keys mentioned to the secret.

#### Tyk Enterprise Developer Portal AWS S3 Secret Access Key

In order to refer Tyk Enterprise Developer Portal connection string to the selected database through Kubernetes secret,
please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 
`DevPortalAwsSecretAccessKey`.

> > [!WARNING]

Author: Use warning shortcode here

> If `tyk-dev-portal.useSecretName` is in use, please add all keys mentioned to the secret.

#### Redis Password

Redis password can also be provided via a secret. Store Redis password in Kubernetes secret and refer to this secret
via `global.redis.passSecret.name` and `global.redis.passSecret.keyName` field, as follows:

```yaml
global:  
  redis:
     passSecret:
       name: "yourSecret"
       keyName: "redisPassKey"

> ```

Reviewer: user just need to refer to the secret as shown in the yaml example

> #### MongoDB or Postgres connection strings

Author: ```suggestion
#### MongoDB or PostgreSQL connection strings
```

> Storage connection strings can also be provided via a secret. Store the connection string in Kubernetes secret and 

Author: _Storage connection strings can also be provided via a secret. Store the connection string in a Kubernetes secret and  refer to this secret via `global.{mongo,postgres}.connectionURLSecret.name` and `global.{mongo,postgres}.connectionURLSecret.keyName` field, as follows:_

refer to this secret via `global.{mongo,postgres}.connectionURLSecret.name` and `global.{mongo,postgres}.connectionURLSecret.keyName` field, 
as follows:

- MongoDB:
```yaml
global:  
  mongo:
    connectionURLSecret:
       name: "yourSecret"
       keyName: "redisPassKey"
```

- Postgres:
```yaml
global:
  postgres:
    connectionURLSecret:
       name: "yourSecret"
       keyName: "redisPassKey"
```

### Gateway Configurations

> Configure below inside `tyk-gateway` section.

Author: Refactor this to introduce what the section shows, something like:

This section explains how to configure the ty-gateway section for updating the Gateway version, enabling TLS, enabling autoscaling etc. etc.

Reviewer: updated

#### Update Tyk Gateway Version

> Set version of gateway at `tyk-gateway.gateway.image.tag`. You can find the list of version tags available from [Docker hub](https://hub.docker.com/u/tykio). Please check [Tyk Release notes]({{<ref "/release-notes">}}) carefully while upgrading or downgrading.

Author: Use specific link to Docker Gateway tags and add link to Gateway Release Notes overview in product stack section

Reviewer: updated

#### Enabling TLS

*Enable TLS*

We have provided an easy way to enable TLS via the `global.tls.gateway` flag. Setting this value to true will
automatically enable TLS using the certificate provided under tyk-gateway/certs/.

*Configure TLS secret*

If you want to use your own key/cert pair, please follow the following steps:
1. Create a TLS secret using your cert and key pair.
2. Set `global.tls.gateway` to true.
3. Set `tyk-gateway.gateway.tls.useDefaultTykCertificate` to false.
4. Set `tyk-gateway.gateway.tls.secretName` to the name of the newly created secret.

*Add Custom Certificates*

To add your custom Certificate Authority(CA) to your containers, you can mount your CA certificate directly into /etc/ssl/certs folder.

```yaml
   extraVolumes: 
     - name: self-signed-ca
       secret:
         secretName: self-signed-ca-secret
   extraVolumeMounts: 
     - name: self-signed-ca
       mountPath: "/etc/ssl/certs/myCA.pem"
       subPath: myCA.pem
```

#### Enabling gateway autoscaling

> You can enable autoscaling of the gateway by `--set tyk-gateway.gateway.autoscaling.enabled=true`. By default, it will enable `Horizontal Pod Autoscaler` resource with target average CPU utilisation at 60%, scaling between 1 and 3 instances. To customize those values you can modify below section of `values.yaml`:

Author: ```suggestion
You can enable autoscaling of the gateway by `--set tyk-gateway.gateway.autoscaling.enabled=true`. By default, it will enable the `Horizontal Pod Autoscaler` resource with target average CPU utilisation at 60%, scaling between 1 and 3 instances. To customise those values you can modify the tyk-gateway section of the `values.yaml` file as shown below:
```

      enabled: true

      # specify ingress controller class name
      className: ""

      # annotations for ingress
      annotations: {}

      # ingress rules
      hosts:
        - host: tyk-gw.local
          paths:
            - path: /
              pathType: ImplementationSpecific

      # tls configuration for ingress
      #  - secretName: chart-example-tls
      #    hosts:
      #      - chart-example.local
      tls: []
```

*Control Port*

Set `tyk-gateway.gateway.control.enabled` to true will allow you to run the [Gateway API]({{<ref "/tyk-gateway-api">}}) on a separate port and protect it behind a firewall if needed.

#### Sharding

Configure the gateways to load APIs with specific tags only by enabling `tyk-gateway.gateway.sharding.enabled`, and set `tags` to comma separated lists of matching tags.

```yaml
    # Sharding gateway allows you to selectively load APIs to specific gateways.
    # If enabled make sure you have at least one gateway that is not sharded.
    # Also be sure to match API segmentation tags with the tags selected below.
    sharding:
      enabled: true
      tags: "edge,dc1,product"
```

#### Deploy additional gateway groups

> `tyk-stack` chart manages one Gateway Deployment in the same namespace as Tyk Dashboard. You can flexibly deploy additional gateways using `tyk-gateway` component chart. With gateway sharding, it is useful for:

Author: ```suggestion
The `tyk-stack` chart manages one Gateway deployment in the same namespace as Tyk Dashboard. You can flexibly deploy additional gateways using the `tyk-gateway` component chart. With gateway sharding, it is useful for:
```

> - Deploy Gateways in different networks,

Author: ```suggestion
- Deploying Gateways in different networks.
```

> - Deploy Gateways with different resources and autoscaling parameters,

Author: ```suggestion
- Deploying Gateways with different resources and autoscaling parameters.
```

  sharding:
    enabled: true
    tags: "gw-dmz"

  ...

  # analyticsEnabled property is used to enable/disable analytics.
  # If set to empty or nil, analytics will be enabled/disabled based on `global.components.pump`.
  analyticsEnabled: "true"

  # used to decide whether to send the results back directly to Tyk without a hybrid pump
  # if you want to send analytics to control plane instead of pump, change analyticsConfigType to "rpc"
  analyticsConfigType: ""
```

Run the following command to deploy additional Gateways in namespace `another-namespace`.
```bash
helm install another-gateway tyk-helm/tyk-gateway --namespace another-namespace -f values.yaml
```

#### OpenTelemetry
To enable OpenTelemetry for Gateway set `gateway.opentelemetry.enabled` flag to true. It is disabled by default.

You can also configure connection settings for it's exporter. By default `grpc` exporter is enabled on `localhost:4317` endpoint.

 To enable TLS settings for the exporter, you can set `gateway.opentelemetry.tls.enabled` to true. 

### Pump Configurations

To enable Pump, set `global.components.pump` to true, and configure below inside `tyk-pump` section.

> <!-- BEGIN import from pump doc -->

Author: Remember to remove this comment

        # it is disabled.
        # Tyk Pump can be monitored with PodMonitor Custom Resource of Prometheus Operator.
        # If enabled, PodMonitor resource is created based on .Values.pump.prometheusPump.prometheusOperator.podMonitorSelector
        # for Tyk Pump.
        enabled: false
        # podMonitorSelector represents a podMonitorSelector of your Prometheus resource. So that
        # your Prometheus resource can select PodMonitor objects based on selector defined here.
        # Please set this field to the podMonitorSelector field of your monitoring.coreos.com/v1
        # Prometheus resource's spec.
        #
        # You can check the podMonitorSelector via:
        #   kubectl describe prometheuses.monitoring.coreos.com <PROMETHEUS_POD>
        podMonitorSelector:
          release: prometheus-stack
```

#### Mongo pump
If you are using the MongoDB pumps in the tyk-oss installation you will require MongoDB installed for that as well.

To install Mongo you can use these rather excellent charts provided by Bitnami:

```bash
helm install tyk-mongo bitnami/mongodb --version {HELM_CHART_VERSION} --set "replicaSet.enabled=true" -n tyk
```

> (follow notes from the installation output to get connection details and update them in `values.yaml` file)

Author: refactor so this is a sentence rather than bracketed note

NOTE: [Here is](https://tyk.io/docs/planning-for-production/database-settings/) list of supported MongoDB versions. Please make sure you are installing mongo helm chart that matches these versions.

> *Important Note regarding MongoDB:* This helm chart enables the PodDisruptionBudget for MongoDB with an arbiter replica-count of 1. If you intend to perform system maintenance on the node where the MongoDB pod is running and this maintenance requires for the node to be drained, this action will be prevented due the replica count being 1. Increase the replica count in the helm chart deployment to a minimum of 2 to remedy this issue.

Author: Use the warning shortcode here

```yaml
 # Set mongo connection details if you want to configure mongo pump.     
 mongo:
    # The mongoURL value will allow you to set your MongoDB address.
    # Default value: mongodb://mongo.{{ .Release.Namespace }}.svc.cluster.local:27017/tyk_analytics
    # mongoURL: mongodb://mongo.tyk.svc.cluster.local:27017/tyk_analytics
    # If your MongoDB has a password you can add the username and password to the url
    # mongoURL: mongodb://root:pass@tyk-mongo-mongodb.tyk.svc.cluster.local:27017/tyk_analytics?authSource=admin
    mongoURL: <MongoDB address>

    # Enables SSL for MongoDB connection. MongoDB instance will have to support that.
    # Default value: false
    # useSSL: false
```

#### SQL pump
If you are using the SQL pumps in the tyk-oss installation you will require PostgreSQL installed for that as well.

To install PostgreSQL you can use these rather excellent charts provided by Bitnami:

```bash
helm install tyk-postgres bitnami/postgresql --set "auth.database=tyk_analytics" -n tyk
```

> (follow notes from the installation output to get connection details and update them in `values.yaml` file)

Author: Refactor this to remove brackets and replace with sentence

```yaml
  # Postgres connection string parameters.
  postgres:
    # host corresponds to the host name of postgres
    host: tyk-postgres-postgresql.tyk.svc
    # port corresponds to the port of postgres
    port: 5432
    # user corresponds to the user of postgres
    user: postgres
    # password corresponds to the password of the given postgres user in selected database
    password:
    # database corresponds to the database to be used in postgres
    database: tyk_analytics
    # sslmode corresponds to if postgres runs in sslmode (https)
    sslmode: disable
    # Connection string can also be set using a secret. Provide the name of the secret and key below.
    # connectionStringSecret:
    #   name: ""
    #   keyName: ""
```

#### Uptime Pump
Uptime Pump can be configured by setting `pump.uptimePumpBackend` in values.yaml file. It support following values

> 1. mongo: Used to set mongo pump for uptime analytics. Mongo Pump should be enabled.

Author: Capitalise mongo pump

2. postgres: Used to set postgres pump for uptime analytics. Postgres Pump should be enabled.
3. empty: Used to disable uptime analytics.

```yaml
    # uptimePumpBackend configures uptime Tyk Pump. ["", "mongo", "postgres"].
    # Set it to "" for disabling uptime Tyk Pump. By default, uptime pump is disabled.
    uptimePumpBackend: ""
```

#### Other Pumps
To setup other backends for pump, refer to this [document](https://github.com/TykTechnologies/tyk-pump/blob/master/README.md#pumps--back-ends-supported) and add the required environment variables in `pump.extraEnvs`

### Tyk Dashboard

#### Tyk Dashboard License (Required)

Tyk Dashboard License is required. It can be set up in `global.license.dashboard` or through secret `global.secrets.useSecretName`. The secret should contain a key called DashLicense.

```yaml
global:
  license:
    # The license key needed by Tyk Dashboard to work.
    #
    # NOTE: If you do not want to store license as a plain text in the file, you can use a Kubernetes secret
    # that stores the dashboard license. Please see `.global.secrets.useSecretName`.
    dashboard: ""
```

### Tyk Bootstrap

> To enable bootstrapping, set `global.components.bootstrap` to `true`. It would run [tyk-k8s-bootstrap](https://github.com/TykTechnologies/tyk-k8s-bootstrap) to bootstrap `tyk-stack` and to create Kubernetes secrets that can be utilized in Tyk Operator and Tyk Enterprise Developer Portal.

Author: ```suggestion
To enable bootstrapping, set `global.components.bootstrap` to `true`. It would run [tyk-k8s-bootstrap](https://github.com/TykTechnologies/tyk-k8s-bootstrap) to bootstrap `tyk-stack` and to create Kubernetes secrets that can be utilised in Tyk Operator and Tyk Enterprise Developer Portal.
```

#### Bootstrapped Environments

> If Tyk is already bootstrapped, the application will bypass the creation of the Tyk Organization and Admin User, proceeding directly with the creation of Kubernetes Secrets.

Author: ```suggestion
If Tyk is already bootstrapped, the application will bypass the creation of the Tyk Organisation and Admin User, proceeding directly with the creation of Kubernetes Secrets.
```

Ensure that these environment variables are set appropriately to `postInstall` hook for bootstrapped environments.

### Tyk Enterprise Developer Portal Configurations

To enable Tyk Enterprise Developer Portal, set `global.components.devPortal` to true, and configure below inside `tyk-dev-portal` section.

#### Tyk Enterprise Developer Portal License (Required)

Tyk Enterprise Developer Portal License is required. It can be set up in `tyk-dev-portal.license` or through secret `global.secrets.useSecretName`. The secret should contain a key called `DevPortalLicense`.

```yaml
tyk-dev-portal:
  # Tyk Developer Portal license.
  license: ""
```

#### Tyk Enterprise Developer Portal Database

By default, Tyk Enterprise Developer Portal use `sqlite3` to store portal metadata. If you want to use other SQL Database, please modify the section below.

```yaml
tyk-dev-portal:
  database:
    # This selects the SQL dialect to be used
    # The supported values are mysql, postgres and sqlite3
    dialect: "sqlite3"
    connectionString: "db/portal.db"
    enableLogs: false
    maxRetries: 3
    retryDelay: 5000
```

#### Storage Settings

> Tyk Enterprise Developer Portal supports different storage options for storing the portal's CMS assets such as images, theme files and Open API Specification files. Please see the [Enterprise Developer Portal Storage settings]({{<ref "product-stack/tyk-enterprise-developer-portal/deploy/configuration#portal-settings">}}) page for all the available options. 

Author: ```suggestion
Tyk Enterprise Developer Portal supports different storage options for storing the Portal's CMS assets such as images, theme files and Open API Specification files. Please see the [Enterprise Developer Portal Storage settings]({{<ref "product-stack/tyk-enterprise-developer-portal/deploy/configuration#portal-settings">}}) page for all the available options. 
```

If you use the file system as storage, please set `tyk-dev-portal.storage.type` to `fs`, and configure `tyk-dev-portal.storage.persistence` to mount an existing persistent volume to Tyk Enterprise Developer Portal.

If you use [AWS S3](https://aws.amazon.com/s3/) as storage, please set `tyk-dev-portal.storage.type` to `s3`, and configure `tyk-dev-portal.storage.s3` section with credentials to access AWS S3 bucket.

> If you use database as storage, please set `tyk-dev-portal.storage.type` to `db`, and configure `tyk-dev-portal.database` section with database connection details.

Author: ```suggestion
If you use a database as storage, please set `tyk-dev-portal.storage.type` to `db`, and configure the `tyk-dev-portal.database` section with database connection details.
```
